source/grid/manifold_lib.cc

```diff
@@ -549,6 +549,127 @@ TorusManifold<dim>::push_forward_gradient(const Point<3> &chart_point) const
+{
+  for (unsigned int v=0; v<GeometryInfo<dim>::vertices_per_cell; ++v)
+    surrounding_points[v] = cell->vertex(v);
+  for (unsigned int f=0; f<GeometryInfo<dim>::faces_per_cell; ++f)
+    face_manifolds[f] = &cell->face(f)->get_manifold();
+}
+
+
+
+template <int dim, int spacedim>
+Point<spacedim>
+TransfiniteInterpolationManifold<dim,spacedim>
+::push_forward(const Point<dim> &chart_point) const
+{
+  // check that the point is in the unit interval which is the chart that is
+  // prescribed
+  for (unsigned int d=0; d<dim; ++d)
+    Assert(chart_point[d] >= -1e-7 && chart_point[d] <= 1.+1e-7,
+           ExcMessage("chart_point is not in unit interval for component "
+                      + std::to_string(d)));
+
+  // formula see wikipedia
+  // https://en.wikipedia.org/wiki/Transfinite_interpolation
+  // S(u,v) = (1-v)c_1(u)+v c_3(u) + (1-u)c_2(v) + u c_4(v) -
+  //   [(1-u)(1-v)P_1 + u(1-v) P_1 + (1-u)v P_2 + uv P_3]
+
+  Point<spacedim> new_point;
+  for (unsigned int v=0; v<GeometryInfo<dim>::vertices_per_cell; ++v)
+    new_point -= GeometryInfo<dim>::d_linear_shape_function(chart_point, v) *
+                 surrounding_points[v];
+
+  std::vector<double> weights(GeometryInfo<dim>::vertices_per_face);
+  std::vector<Point<spacedim> > points(GeometryInfo<dim>::vertices_per_face);
+  for (unsigned int face=0; face<GeometryInfo<dim>::faces_per_cell; ++face)
+    {
+      // set the point on the current face
+      Point<dim> point_copy = chart_point;
+      point_copy[face/2] = face%2;
+
+      for (unsigned int v=0; v<GeometryInfo<dim>::vertices_per_face; ++v)
+        {
+          const unsigned int index = GeometryInfo<dim>::face_to_cell_vertices(face,v);
+          points[v] = surrounding_points[index];
+          weights[v] = GeometryInfo<dim>::d_linear_shape_function (point_copy,
+                                                                   index);
+        }
+      const double my_weight = face%2 ? chart_point[face/2] : 1-chart_point[face/2];
+      new_point += my_weight * face_manifolds[face]->get_new_point(points, weights);
+    }
+  return new_point;
+}
+
+
+
+template <int dim, int spacedim>
+DerivativeForm<1,dim,spacedim>
+TransfiniteInterpolationManifold<dim,spacedim>
+::push_forward_gradient(const Point<dim> &chart_point) const
+{
+  // compute the derivative with the help of finite differences
+  const double eps = 1e-8;
+  Point<spacedim> point = push_forward(chart_point);
+  DerivativeForm<1,dim,spacedim> grad;
+  for (unsigned int d=0; d<dim; ++d)
+    {
+      Point<dim> modified = chart_point;
+
+      // avoid checking outside of the unit interval
+      if (modified[d] > 0.5)
+        modified[d] -= eps;
+      else
+        modified[d] += eps;
+      Tensor<1,spacedim> difference = push_forward(modified) - point;
+      for (unsigned int e=0; e<spacedim; ++e)
+        grad[e][d] = difference[e] / (modified[d] - chart_point[d]);
```

Comment:
not that it will make much of a difference, but you know what `modified[d] - chart_point[d]` is (namely +eps or -eps). It may be easier to read if you used it that way.